src/polymer/content-to-slot-usages.ts

```diff
@@ -172,27 +121,219 @@ class ContentToSlot extends HtmlRule {
+ *
+ * This is intended for handling only selectors that match an individual element
+ * in isolation, it does throws if the selector talks about relationships
+ * between elements like `.foo .bar` or `.foo > .bar`.
+ */
+function elementSelectorToPredicate(simpleSelector: string): dom5.Predicate {
+  const parsed = cssWhat(simpleSelector);
+  // The output of cssWhat is two levels of arrays. The outer level are any
+  // selectors joined with a comma, so it matches if any of the inner selectors
+  // match. The inner array are simple selectors like `.foo` and `#bar` which
+  // must all match.
+  return dom5.predicates.OR(...parsed.map((simpleSelectors) => {
+    return dom5.predicates.AND(
+        ...simpleSelectors.map(simpleSelectorToPredicate));
+  }));
+}
+
+function simpleSelectorToPredicate(selector: cssWhat.Simple) {
+  switch (selector.type) {
+    case 'adjacent':
+    case 'child':
+    case 'descendant':
+    case 'parent':
+    case 'sibling':
+    case 'pseudo':
+      throw new Error(`Unsupported CSS operator: ${selector.type}`);
+    case 'attribute':
+      return attributeSelectorToPredicate(selector);
+    case 'tag':
+      return dom5.predicates.hasTagName(selector.name);
+    case 'universal':
+      return () => true;
+  }
+  const never: never = selector;
+  throw new Error(`Unexpected node type from css parser: ${never}`);
+}
+
+function attributeSelectorToPredicate(selector: cssWhat.Attribute):
+    dom5.Predicate {
+  switch (selector.action) {
+    case 'exists':
+      return dom5.predicates.hasAttr(selector.name);
+    case 'equals':
+      return dom5.predicates.hasAttrValue(selector.name, selector.value);
+    case 'start':
+      return (el) => {
+        const attrValue = dom5.getAttribute(el, selector.name);
+        return attrValue != null && attrValue.startsWith(selector.value);
+      };
+    case 'end':
+      return (el) => {
+        const attrValue = dom5.getAttribute(el, selector.name);
+        return attrValue != null && attrValue.endsWith(selector.value);
+      };
+    case 'element':
+      return dom5.predicates.hasSpaceSeparatedAttrValue(
+          selector.name, selector.value);
+    case 'any':
+      return (el) => {
+        const attrValue = dom5.getAttribute(el, selector.name);
+        return attrValue != null && attrValue.includes(selector.value);
+      };
+  }
+  const never: never = selector.action;
+  throw new Error(
+      `Unexpected type of attribute matcher from CSS parser ${never}`);
+}
+
+const staticConfig =
+    new Map<string, Array<{predicate: dom5.Predicate, slot: string}>>();
+
+// Configure statically known slot->content conversions.
+function addPredicate(
+    tagname: string, slots: Array<{selector: string, slot: string}>) {
+  staticConfig.set(
+      tagname, slots.map((s) => ({
+                           predicate: elementSelectorToPredicate(s.selector),
+                           slot: s.slot
+                         })));
 }
 
+addPredicate(
```

Comment:
Probably material for a separate PR, but I'd define these predicates externally (e.g. a json file) and make the path configurable/extensible. This would allow element authors to share their upgrades to users of their elements.

Reply:
I'd prefer to instead make changes to our elements to add the `old-content-selector` attribute and to remove this hard coded set. I added it in while I was initially hacking on this stuff. `old-content-selector` is better in every way.

I'll do that in a separate PR.